File: src/ipc/events.ts
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import type {
  BroadcastStatePayload,
  BroadcastTickPayload,
  ErrorPayload,
  FocusedWindowChangedPayload,
  StartupFlowStatePayload,
  UpdateProgressPayload,
  UpdateStatePayload,
  WindowsChangedPayload,
} from "../types";

export const EVT_WINDOWS_CHANGED = "windows-changed";
export const EVT_BROADCAST_STATE = "broadcast-state-changed";
export const EVT_BROADCAST_TICK = "broadcast-tick";
export const EVT_ERROR = "error";
export const EVT_OPEN_SETTINGS = "open-settings";
export const EVT_PREFS_CHANGED = "prefs-changed";
export const EVT_FOCUSED_WINDOW_CHANGED = "focused-window-changed";
export const EVT_UPDATE_STATE = "update-state-changed";
export const EVT_UPDATE_PROGRESS = "update-progress";
export const EVT_STARTUP_FLOW_STATE = "startup-flow-state";

export const onWindowsChanged = (cb: (p: WindowsChangedPayload) => void): Promise<UnlistenFn> =>
  listen<WindowsChangedPayload>(EVT_WINDOWS_CHANGED, (e) => cb(e.payload));

export const onFocusedWindowChanged = (
  cb: (p: FocusedWindowChangedPayload) => void,
): Promise<UnlistenFn> =>
  listen<FocusedWindowChangedPayload>(EVT_FOCUSED_WINDOW_CHANGED, (e) => cb(e.payload));

export const onBroadcastState = (cb: (p: BroadcastStatePayload) => void): Promise<UnlistenFn> =>
  listen<BroadcastStatePayload>(EVT_BROADCAST_STATE, (e) => cb(e.payload));

export const onBroadcastTick = (cb: (p: BroadcastTickPayload) => void): Promise<UnlistenFn> =>
  listen<BroadcastTickPayload>(EVT_BROADCAST_TICK, (e) => cb(e.payload));

export const onError = (cb: (p: ErrorPayload) => void): Promise<UnlistenFn> =>
  listen<ErrorPayload>(EVT_ERROR, (e) => cb(e.payload));

export const onOpenSettings = (cb: () => void): Promise<UnlistenFn> =>
  listen<void>(EVT_OPEN_SETTINGS, () => cb());

export const onPrefsChanged = (cb: () => void): Promise<UnlistenFn> =>
  listen<void>(EVT_PREFS_CHANGED, () => cb());

export const onUpdateState = (cb: (p: UpdateStatePayload) => void): Promise<UnlistenFn> =>
  listen<UpdateStatePayload>(EVT_UPDATE_STATE, (e) => cb(e.payload));

export const onUpdateProgress = (cb: (p: UpdateProgressPayload) => void): Promise<UnlistenFn> =>
  listen<UpdateProgressPayload>(EVT_UPDATE_PROGRESS, (e) => cb(e.payload));

export const onStartupFlowState = (cb: (p: StartupFlowStatePayload) => void): Promise<UnlistenFn> =>
  listen<StartupFlowStatePayload>(EVT_STARTUP_FLOW_STATE, (e) => cb(e.payload));

File: src-tauri/src/events.rs
use serde::Serialize;

use crate::state::{BroadcastReason, WindowEntry};

pub const EVT_WINDOWS_CHANGED: &str = "windows-changed";
pub const EVT_BROADCAST_STATE: &str = "broadcast-state-changed";
pub const EVT_BROADCAST_TICK: &str = "broadcast-tick";
pub const EVT_ERROR: &str = "error";
pub const EVT_OPEN_SETTINGS: &str = "open-settings";
pub const EVT_PREFS_CHANGED: &str = "prefs-changed";
pub const EVT_FOCUSED_WINDOW_CHANGED: &str = "focused-window-changed";
pub const EVT_UPDATE_STATE: &str = "update-state-changed";
pub const EVT_UPDATE_PROGRESS: &str = "update-progress";
pub const EVT_STARTUP_FLOW_STATE: &str = "startup-flow-state";

#[derive(Debug, Clone, Serialize)]
pub struct WindowsChangedPayload {
    pub windows: Vec<WindowEntry>,
}

#[derive(Debug, Clone, Serialize)]
pub struct FocusedWindowChangedPayload {
    /// HWND of the foreground window if it matches a tracked Dofus window,
    /// otherwise `None`.
    pub focused_hwnd: Option<isize>,
}

#[derive(Debug, Clone, Serialize)]
pub struct BroadcastStatePayload {
    pub enabled: bool,
    pub reason: BroadcastReason,
}

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "kind", content = "data")]
pub enum BroadcastTickPayload {
    Started { followers: usize },
    Finished { ok: usize, failed: usize },
}

#[derive(Debug, Clone, Serialize)]
pub struct ErrorPayload {
    pub message: String,
    pub context: Option<String>,
}

/// Emitted to the frontend via `EVT_UPDATE_STATE`. The TS store has an
/// extra `"idle"` initial state; Rust never emits it (the backend only
/// reports transitions). Keep the lowercase serde tags in sync with the
/// `UpdateState` union in `src/types.ts`.
#[derive(Debug, Clone, Copy, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum UpdateState {
    Checking,
    Available,
    NoUpdate,
    Downloading,
    Installing,
    Error,
}

#[derive(Debug, Clone, Serialize)]
pub struct UpdateStatePayload {
    pub state: UpdateState,
    pub version: Option<String>,
    pub notes: Option<String>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Copy, Serialize)]
pub struct UpdateProgressPayload {
    pub downloaded: u64,
    pub total: Option<u64>,
}
